docs/input/docs/usage/msbuild-task.md

```diff
@@ -84,44 +84,63 @@ appended to it.
 #### Other injected Variables
 
 All other [variables](../more-info/variables) will be injected into an
-internal static class:
+internal static class part of the global namespace similar to this:
```

Comment:
I was just thinking, if `GitversionInformation` is moved to the global namespace, won't this make ILMerging difficult? If several of the merged assemblies are versioned with GitVersion, we would potentially end up with several colliding `GitversionInformation` classes?

Reply:
That poses no problem because the generated classes are, and this is very important,  **internal**. An external assembly cannot see these class unless you use reflection, and when using reflection they are resolved with their full name (which happens to be only the class name) against a specific assembly, therefore, no collision.

This trick, I often use with Visual Studio's "shared projects": they are just bunches of code that get "included" into the consuming project. I always make sure classes in these shared projects are internal so that they can be used from any assembly even if these assemblies reference one another.

Reply:
And if this were an issue, it would already have shown with C# or F# generated classes that are already part of the global namespace (and internal).

Reply:
Mmmh, maybe I spoke too quickly, I missed the bit about ILMerge... Everything being part of the same aszembly in the end, it may be an issue. I'll give it a try, and sorry if you thought I was teaching you how visibility worked in .NET!

Reply:
Well, after a quick googling, it appears this indeeds poses problems: 

> Suppose class A is internal to App, and also to one of the lib (with same namespace). Before merging there will not be any issue. After merging it will become issue to resolve ambiguous reference.

See <https://stackoverflow.com/a/14042227/107552>

Reply:
I see. That's an issue worth pursuingthen, because it's difficult to resolve for consumers of GitVersion. Ideas?

Reply:
I experimented with IL packing using [ILRepack](https://github.com/ravibpatel/ILRepack.Lib.MSBuild.Task). The tool is rather clever: upon detection of the same type in several input assemblies, it keeps one as is (the first iot encounters I suppose) then renames the other ones:

![image](https://user-images.githubusercontent.com/79333/84525261-a4262500-acdb-11ea-98dd-590f3dd368e9.png)

Then I suppose, at least with this tool, your concern with type collision is mitigated. 
I may experiment too with [ILMerge](https://github.com/dotnet/ILMerge) or [Fody](https://github.com/Fody/Fody).

PS: As a refinement, I also looked for a way to tell **ILRepack** to favor a specific source assembly when determining what type to keep. My use case would be: merge plenty of .DLL along the main executable, and be sure the `GitVersionInformation` that remains accessible is the one in the exe. Well, I managed to achieve this (though Idon't know how robust it is) by playing with the order in which assemblies to merge are listed: the last assembly wins; therefore having the main executable merged last seems to guarantee its `GitVersionInformation` is the one that keeps its name.

Reply:
[ILMerge](https://github.com/dotnet/ILMerge) is similar to **ILRepack** given you pass it the correct options. Concerning the type collisions, it is slightly better as the types are renamed according to a preditable scheme:

![image](https://user-images.githubusercontent.com/79333/84569213-1a329680-ad85-11ea-8035-4c584bb7216e.png)

Reply:
And with [Fody + Costura](https://github.com/Fody/Costura), everything just works out of the box. This is because, the principle is radically different. The assembies' IL is not merged, instead the dependent assemblies are embedded as resources in the main assembly, then extracted and loaded at runtime (during module initialization). Thus, the assemblies never cease to exist as such and type resolution just works.

Reply:
If you are interested in witnessing the tests I did, everything is in this repository: <https://github.com/odalet/GitVersionTests>

Reply:
And now for ideas on how this whole "code generation" thing could be tweaked:

* The generated class could be put in a namespace or have a name that will guarantee non-collision. To guarantee this, the class name should be based on the assembly name (and not on some 'default' or root namespace). The class could be inside a namespace named as the assembly or be an inner class inside a class named like the assembly or be a concatenation of assembly name and `GitVersionInformation`...

* A property / command line argument could be provided to let the user choose the name of the generated class.

* I recently stumbled upon the [GitInfo](https://github.com/kzu/GitInfo). Its purpose is similar to **GitVersion**'s with the following differences:
  * Completely MSBuild-based, no C# code at all
  * Seems to have far less feature. It seems to simply expose Git information and does not attempt clever inference of versions.
  * There is however one feature that may be interesting here:

It, too, generates code containing version information. However:
* The version information is made available through constants, not properties or fields.
* The file containing this information is generated soon enough for it to be available at "design-time": it can be used without resorting to reflection.

The advantage of this strategy is that because constants are inlined in the calling code, the original type that exposes them is useless at runtime and therefore, collisions on this type are not anymore an issue. And being able to use the version information without reflection also seems more user-friendly.

So, mixing some of these ideas may be the way to go. However, in any case, it is quite some work, and most probably breaking. In the meantime, my experiments with a selection of assembly merging tools show that the collision problem can be mitigated. 

It's now up to you to reflect upon all this and decide what or what not to do :)

Hope this helped!

Reply:
Constants are a great idea for how we can implement this in a new major version of GitVersion! Whether we'll do this for v6 or not is mainly dependent on timing; if someone submits a pull request implementing this before we go live with v6, it will be released with v6. If not, we would have to postpone it to v7.

Reply:
Generating constants should be rather straightforward. The MSBuild part that generates this soon so that the user can call into it may be a little trickier (I'm not very proficient at MSBuild ;)).

I'll probably come up with an issue that will serve as a discussion and specification of what can be done.